telemetry/data_feature_usage.go

```diff
@@ -0,0 +1,165 @@
+	// init
+	usageInfo := featureUsageInfo{
+		CoprCacheUsed:    make([]*CoprCacheUsedWindowItem, 0, UpdateInterval/time.Hour),
+		ClusterIndexUsed: make(map[string]bool),
+		TiFlashUsed:      make([]*TiFlashUsageItem, 0, UpdateInterval/time.Hour),
+	}
+
+	// coprocessor cache
+	CoprocessorCacheTelemetry.Lock.Lock()
+	maxLen := 0
+	for _, window := range CoprocessorCacheTelemetry.MinuteWindow {
+		timeSince := time.Since(*window.BeginAt)
+		if timeSince >= UpdateInterval {
+			continue
+		}
+		maxLen = mathutil.Max(maxLen, int(timeSince/time.Hour))
+		i := maxLen - int(timeSince/time.Hour)
+		if len(usageInfo.CoprCacheUsed) <= i {
+			usageInfo.CoprCacheUsed = append(usageInfo.CoprCacheUsed, &CoprCacheUsedWindowItem{})
+		}
+		usageInfo.CoprCacheUsed[i].P0.Add(window.P0.Load())
+		usageInfo.CoprCacheUsed[i].P1.Add(window.P1.Load())
+		usageInfo.CoprCacheUsed[i].P10.Add(window.P10.Load())
+		usageInfo.CoprCacheUsed[i].P20.Add(window.P20.Load())
+		usageInfo.CoprCacheUsed[i].P40.Add(window.P40.Load())
+		usageInfo.CoprCacheUsed[i].P80.Add(window.P80.Load())
+		usageInfo.CoprCacheUsed[i].P100.Add(window.P100.Load())
+		if usageInfo.CoprCacheUsed[i].BeginAt == nil {
+			usageInfo.CoprCacheUsed[i].BeginAt = window.BeginAt
+		}
+	}
+	CoprocessorCacheTelemetry.Lock.Unlock()
+
+	// TiFlash usage
+	TiFlashUsageTelemetry.Lock.Lock()
+	maxLen = 0
+	for _, window := range TiFlashUsageTelemetry.MinuteWindow {
+		timeSince := time.Since(*window.BeginAt)
+		if timeSince >= UpdateInterval {
+			continue
+		}
+		maxLen = mathutil.Max(maxLen, int(timeSince/time.Hour))
+		i := maxLen - int(timeSince/time.Hour)
+		if len(usageInfo.TiFlashUsed) <= i {
+			usageInfo.TiFlashUsed = append(usageInfo.TiFlashUsed, &TiFlashUsageItem{})
+		}
+		usageInfo.TiFlashUsed[i].TotalNumbers.Add(window.TotalNumbers.Load())
+		usageInfo.TiFlashUsed[i].TiFlashPushDownNumbers.Add(window.TiFlashPushDownNumbers.Load())
+		usageInfo.TiFlashUsed[i].TiFlashExchangePushDownNumbers.Add(window.TiFlashExchangePushDownNumbers.Load())
+		if usageInfo.TiFlashUsed[i].BeginAt == nil {
+			usageInfo.TiFlashUsed[i].BeginAt = window.BeginAt
+		}
+	}
+	TiFlashUsageTelemetry.Lock.Unlock()
+
+	// cluster index
+	exec := ctx.(sqlexec.RestrictedSQLExecutor)
+	stmt, err := exec.ParseWithParams(context.TODO(), `
+		SELECT sha2(TABLE_NAME, 256) name, TIDB_PK_TYPE
+		FROM information_schema.tables
+		WHERE table_schema not in ('INFORMATION_SCHEMA', 'METRICS_SCHEMA', 'PERFORMANCE_SCHEMA', 'mysql')
+		limit 10000`)
+	if err != nil {
+		return nil, err
+	}
+	rows, _, err := exec.ExecRestrictedStmt(context.TODO(), stmt)
+	if err != nil {
+		return nil, err
+	}
+	for _, row := range rows {
+		if row.Len() < 2 {
+			continue
+		}
+		isClustered := false
+		if row.GetString(1) == "CLUSTERED" {
+			isClustered = true
+		}
+		usageInfo.ClusterIndexUsed[row.GetString(0)] = isClustered
+	}
+
+	// async commit
+	stmt, err = exec.ParseWithParams(context.TODO(), `show config where name = 'storage.enable-async-apply-prewrite'`)
```

Comment:
Async-commit is controlled by a session variable: `tidb_enable_async_commit`. You can remove codes about this part. I will add it later.